### PR TITLE
Draco compression support for GLB files

### LINKED ISSUE #2001: Support glTF's KHR_draco_mesh_compression extension
### Description

I think other major WebGL libraries support glTF's `KHR_draco_mesh_compression` extension.
The `Draco` format is an excellent compression format.
https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_draco_mesh_compression
I expect PlayCanvas to support the `Draco` format.

### PR DESCRIPTION
**Pull request by mvaligursky**
Fixes #2001

Draco mesh compression is now supported within GLB files. See library here: https://github.com/google/draco

Draco modules version: 1.3.6

Test stats on Bed Model:

Screen Shot 2020-04-23 at 11 58 11 AM

- vertices: 496495
- triangles: 801397
- uncompressed GLB: 22.8MB (95ms to extract meshes on Macbook Pro)
- compressed GLB: 2MB (990ms to decompress meshes on Macbook Pro)

### CODE REVIEW

**Comment by willeastcott** [comment]
I'd suggest placing the draco WASM module in `examples\assets\lib\draco`. You could also rename the files with a similar naming convention. e.g.:

```
draco.js
draco.wasm.js
draco.wasm.wasm
```

**Comment by mvaligursky** [comment]
I've moved and renamed draco library files as suggested, and addressed lint issues.
stil WIP, looking at supporting various types instead of using float32 for all attribute arrays when decompressing

**Comment by willeastcott** [comment]
Nearly but not quite. 😄  `draco_wasm.js` -> `draco.wasm.js`

File: src/resources/parser/glb-parser.js [hunk 1]
```diff
@@ -98,18 +98,31 @@ Object.assign(pc, function () {
...
-            'TEXCOORD_0': pc.SEMANTIC_TEXCOORD0,
-            'TEXCOORD_1': pc.SEMANTIC_TEXCOORD1
+    var generateNormals = function (sourceDesc, vertexDesc, positions, numVertices, indices) {
+
+        if (!indices) {
+            indices = generateIndices(numVertices);
+        }
+
+        // generate normals
+        var normals = Float32Array.from(pc.calculateNormals(positions, indices));
```
Comment (willeastcott) [inline src/resources/parser/glb-parser.js:108]: Sorry. This doesn't work in IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/from

Comment (mvaligursky) [reply src/resources/parser/glb-parser.js:108]: I'll replace this .. this is existing code in main, so perhaps we wait till fix is in.

Comment (mvaligursky) [reply src/resources/parser/glb-parser.js]: fixed

File: src/resources/parser/glb-parser.js [hunk 2]
```diff
@@ -298,24 +388,120 @@ Object.assign(pc, function () {
...
+                                    outputGeometry = new decoderModule.PointCloud();
+                                    status = decoder.DecodeBufferToPointCloud(buffer, outputGeometry);
+                                    break;
+                                case decoderModule.TRIANGULAR_MESH:
+                                    primitiveType = pc.PRIMITIVE_TRIANGLES;
+                                    outputGeometry = new decoderModule.Mesh();
+                                    status = decoder.DecodeBufferToMesh(buffer, outputGeometry);
+                                    break;
+                                default:
+                                case decoderModule.INVALID_GEOMETRY_TYPE:
```
Comment (slimbuck) [inline src/resources/parser/glb-parser.js:445]: would you mind putting ```default:``` last?

File: src/resources/parser/glb-parser.js [hunk 3]
```diff
@@ -298,24 +388,120 @@ Object.assign(pc, function () {
...
+                                    console.error('Mesh asset - invalid draco compressed geometry type: ' + geometryType);
+                                    break;
+                            }
 
-            var mesh = new pc.Mesh();
+                            if (!status.ok() || outputGeometry.ptr == 0) {
+                                // #ifdef DEBUG
+                                console.error("Failed to decode draco compressed asset: " + status.error_msg());
+                                // #endif
```
Comment (slimbuck) [inline src/resources/parser/glb-parser.js:453]: should the draco code below continue to be executed if this fails? shouldn't this result in load failure?

File: src/resources/parser/glb-parser.js [hunk 4]
```diff
@@ -298,24 +388,120 @@ Object.assign(pc, function () {
...
+                    if (decoderModule) {
+                        var extDraco = extensions.KHR_draco_mesh_compression;
+                        if (extDraco.hasOwnProperty('attributes')) {
+                            var bufferView = bufferViews[extDraco.bufferView];
+                            var arrayBuffer = buffers[bufferView.buffer];
+                            var uint8Buffer = new Uint8Array(arrayBuffer.buffer, arrayBuffer.byteOffset + bufferView.byteOffset, bufferView.byteLength);
+                            var buffer = new decoderModule.DecoderBuffer();
+                            buffer.Init(uint8Buffer, uint8Buffer.length);
+
+                            var decoder = new decoderModule.Decoder();
```
Comment (slimbuck) [inline src/resources/parser/glb-parser.js:429]: how slow is it to construct a decoder instance? would there be any benefit in caching it?

File: src/resources/parser/glb-parser.js [hunk 5]
```diff
@@ -327,7 +513,7 @@ Object.assign(pc, function () {
                     // don't support 32bit index data
                     indexFormat = pc.INDEXFORMAT_UINT32;
                 }
-                var numIndices = indices.length;
+                numIndices = indices.length;
```
Comment (slimbuck) [inline src/resources/parser/glb-parser.js:516]: can probably remove numIndices here (or make line 519 also use it).

File: src/resources/parser/glb-parser.js (hunk 2 above)
Comment (mvaligursky) [reply src/resources/parser/glb-parser.js:444]: done

File: src/resources/parser/glb-parser.js (hunk 3 above)
Comment (mvaligursky) [reply src/resources/parser/glb-parser.js]: done

File: src/resources/parser/glb-parser.js (hunk 5 above)
Comment (mvaligursky) [reply src/resources/parser/glb-parser.js]: done

File: src/resources/parser/glb-parser.js [hunk 6]
```diff
@@ -253,6 +207,142 @@ Object.assign(pc, function () {
...
+                        for (k = 0; k < numValues; k++)
+                            values[k] = valuesFloat32[k] * 255;
+                    } else {
+                        decoder.GetAttributeDataArrayForAllPoints(outputGeometry, attribute, decoderModule.DT_UINT8, dataSize, ptr);
+                        values = new Uint8Array(decoderModule.HEAPU8.buffer, ptr, numValues).slice();
+                    }
+                    break;
+
+                default:
+                    console.error("Element type not implemented: " + storageType);
```
Comment (willeastcott) [inline src/resources/parser/glb-parser.js:290]: I'd wrap this (and any console logging) with an #ifdef DEBUG

Comment (mvaligursky) [reply src/resources/parser/glb-parser.js:291]: done

**Comment by marklundin** [comment]
@mvaligursky I didn't realise Draco decompression was already in the engine! So theoretically, if we had a local copy of a project, we could run through all the glb files, compress them with Draco and replace them, then as long as we include the Draco lib in the global object, we can just run the project and it will automatically decompress and use them as normal? 🙏

**Comment by Maksims** [comment]
Does the GLB parsing+decompression happen in a web worker?

**Comment by mvaligursky** [comment]
@marklundin - yes, this should work fine, I don't expect a problem with this. I think to get the loader working, you just need to drop those 3 draco libraries into Editor project, that should work.
https://github.com/playcanvas/engine/tree/master/examples/lib/draco

But as @Maksims mentioned - the decompression takes place on the main thread at the moment, so it costs a bit. See some timing in the description of this PR.

**Comment by marklundin** [comment]
Ok that's good to know! I think in three.js they [fire up a worker](<script src='/decoder/draco.js' />, but I'd be keento know what how performant the decompression is.

Also have you tried with the latest versions? https://github.com/google/draco/tree/master/javascript

**Comment by marklundin** [comment]
on a side note compiling the shaders in a worker would be amazing 😆

**Comment by mvaligursky** [comment]
yeah, there's lots we want to do.
This is on the list as well .. compresses just as well as draco, but decompression is a lot faster: https://github.com/zeux/meshoptimizer
and handles sparse morph targets unlike draco.

**Pull request merged**